File: src/models/Booking.js
const mongoose = require('mongoose');

const BookingSchema = new mongoose.Schema({
  listingId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Listing',
    required: true,
  },
  guestId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  hostId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  checkInDate: {
    type: Date,
    required: [true, 'Please provide check-in date'],
  },
  checkOutDate: {
    type: Date,
    required: [true, 'Please provide check-out date'],
  },
  numberOfGuests: {
    type: Number,
    required: true,
    min: 1,
  },
  pricing: {
    nightlyRate: {
      type: Number,
      required: true,
    },
    numberOfNights: {
      type: Number,
      required: true,
    },
    subtotal: {
      type: Number,
      required: true,
    },
    cleaningFee: {
      type: Number,
      default: 0,
    },
    serviceFee: {
      type: Number,
      default: 0,
    },
    taxes: {
      type: Number,
      default: 0,
    },
    total: {
      type: Number,
      required: true,
    },
    currency: {
      type: String,
      default: 'TZS',
    },
  },
  status: {
    type: String,
    enum: [
      'pending',
      'confirmed',
      'cancelled_by_guest',
      'cancelled_by_host',
      'completed',
      'in_progress',
      'refunded'
    ],
    default: 'pending',
  },
  payment: {
    method: {
      type: String,
      enum: ['zenopay'],
      required: true,
    },
    transactionId: String,
    orderId: String, // ZenoPay order ID
    status: {
      type: String,
      enum: ['pending', 'completed', 'failed', 'refunded'],
      default: 'pending',
    },
    paidAt: Date,
  },
  guestDetails: {
    fullName: {
      type: String,
      required: true,
    },
    phoneNumber: {
      type: String,
      required: true,
    },
    email: {
      type: String,
      required: true,
    },
    numberOfAdults: {
      type: Number,
      default: 1,
    },
    numberOfChildren: {
      type: Number,
      default: 0,
    },
    specialRequests: String,
  },
  cancellation: {
    cancelledBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
    },
    cancelledAt: Date,
    reason: String,
    refundAmount: Number,
  },
  checkInConfirmed: {
    type: Boolean,
    default: false,
  },
  checkOutConfirmed: {
    type: Boolean,
    default: false,
  },
}, {
  timestamps: true,
});

// Indexes
BookingSchema.index({ listingId: 1 });
BookingSchema.index({ guestId: 1 });
BookingSchema.index({ hostId: 1 });
BookingSchema.index({ status: 1 });
BookingSchema.index({ checkInDate: 1 });
BookingSchema.index({ checkOutDate: 1 });
BookingSchema.index({ createdAt: -1 });

// Validate check-out is after check-in
BookingSchema.pre('save', function(next) {
  if (this.checkOutDate <= this.checkInDate) {
    next(new Error('Check-out date must be after check-in date'));
  }
  next();
});

// Calculate number of nights
BookingSchema.pre('save', function(next) {
  const oneDay = 24 * 60 * 60 * 1000;
  this.pricing.numberOfNights = Math.round(
    Math.abs((this.checkOutDate - this.checkInDate) / oneDay)
  );
  next();
});

module.exports = mongoose.model('Booking', BookingSchema);

File: mobile/lib/core/models/booking.dart
class Booking {
  final String id;
  final String listingId;
  final String guestId;
  final String hostId;
  final DateTime checkInDate;
  final DateTime checkOutDate;
  final int numberOfGuests;
  final BookingPricing pricing;
  final String status;
  final PaymentInfo payment;
  final DateTime createdAt;
  final bool checkInConfirmed;
  final BookingPayout? payout;
  final BookingArrival? arrival;
  
  Booking({
    required this.id,
    required this.listingId,
    required this.guestId,
    required this.hostId,
    required this.checkInDate,
    required this.checkOutDate,
    required this.numberOfGuests,
    required this.pricing,
    required this.status,
    required this.payment,
    required this.createdAt,
    required this.checkInConfirmed,
    this.payout,
    this.arrival,
  });
  
  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['_id'] ?? json['id'],
      listingId: json['listingId'] is String ? json['listingId'] : json['listingId']['_id'],
      guestId: json['guestId'] is String ? json['guestId'] : json['guestId']['_id'],
      hostId: json['hostId'] is String ? json['hostId'] : json['hostId']['_id'],
      checkInDate: DateTime.parse(json['checkInDate']),
      checkOutDate: DateTime.parse(json['checkOutDate']),
      numberOfGuests: json['numberOfGuests'],
      pricing: BookingPricing.fromJson(json['pricing']),
      status: json['status'],
      payment: PaymentInfo.fromJson(json['payment']),
      createdAt: DateTime.parse(json['createdAt']),
      checkInConfirmed: json['checkInConfirmed'] ?? false,
      payout: json['payout'] != null ? BookingPayout.fromJson(json['payout']) : null,
      arrival: json['arrival'] != null ? BookingArrival.fromJson(json['arrival']) : null,
    );
  }
  
  int get numberOfNights => checkOutDate.difference(checkInDate).inDays;

  bool get requiresArrivalConfirmation {
    if (checkInConfirmed) return false;
    if (arrival == null) return false;
    return arrival!.requiresConfirmation ?? false;
  }
}

class BookingPricing {
  final double nightlyRate;
  final int numberOfNights;
  final double subtotal;
  final double cleaningFee;
  final double serviceFee;
  final double taxes;
  final double total;
  final String currency;
  
  BookingPricing({
    required this.nightlyRate,
    required this.numberOfNights,
    required this.subtotal,
    required this.cleaningFee,
    required this.serviceFee,
    required this.taxes,
    required this.total,
    required this.currency,
  });
  
  factory BookingPricing.fromJson(Map<String, dynamic> json) {
    return BookingPricing(
      nightlyRate: (json['nightlyRate'] ?? 0).toDouble(),
      numberOfNights: json['numberOfNights'] ?? 0,
      subtotal: (json['subtotal'] ?? 0).toDouble(),
      cleaningFee: (json['cleaningFee'] ?? 0).toDouble(),
      serviceFee: (json['serviceFee'] ?? 0).toDouble(),
      taxes: (json['taxes'] ?? 0).toDouble(),
      total: (json['total'] ?? 0).toDouble(),
      currency: json['currency'] ?? 'TZS',
    );
  }
}

class PaymentInfo {
  final String method;
  final String status;
  final String? transactionId;
  final DateTime? paidAt;
  
  PaymentInfo({
    required this.method,
    required this.status,
    this.transactionId,
    this.paidAt,
  });
  
  factory PaymentInfo.fromJson(Map<String, dynamic> json) {
    return PaymentInfo(
      method: json['method'],
      status: json['status'],
      transactionId: json['transactionId'],
      paidAt: json['paidAt'] != null ? DateTime.parse(json['paidAt']) : null,
    );
  }
}

class BookingPayout {
  final String status;
  final double hostAmount;
  final double platformFee;
  final String currency;
  final DateTime? autoReleaseAt;

  BookingPayout({
    required this.status,
    required this.hostAmount,
    required this.platformFee,
    required this.currency,
    this.autoReleaseAt,
  });

  factory BookingPayout.fromJson(Map<String, dynamic> json) {
    return BookingPayout(
      status: json['status'] ?? 'pending',
      hostAmount: (json['hostAmount'] ?? 0).toDouble(),
      platformFee: (json['platformFee'] ?? 0).toDouble(),
      currency: json['currency'] ?? 'TZS',
      autoReleaseAt: json['autoReleaseAt'] != null ? DateTime.parse(json['autoReleaseAt']) : null,
    );
  }
}

class BookingArrival {
  final bool? requiresConfirmation;
  final DateTime? confirmedAt;
  final DateTime? autoConfirmedAt;

  BookingArrival({
    this.requiresConfirmation,
    this.confirmedAt,
    this.autoConfirmedAt,
  });

  factory BookingArrival.fromJson(Map<String, dynamic> json) {
    return BookingArrival(
      requiresConfirmation: json['requiresConfirmation'],
      confirmedAt: json['confirmedAt'] != null ? DateTime.parse(json['confirmedAt']) : null,
      autoConfirmedAt: json['autoConfirmedAt'] != null ? DateTime.parse(json['autoConfirmedAt']) : null,
    );
  }
}
























